ci: Automate generation and update of docs-builder image

qmonnet opened the pull request:
The docs-builder image is used for the CI wofkflow for testing documentation. Each time we change the list of requirements (Sphinx dependencies etc.) or the Dockerfile for that image, we have to generate a new image to use in the CI workflow.

So far these images have been generated and pushed to Docker manually (thanks Joe!). But this is not fun or practical [citation needed]. This commit introduces a new CI workflow to generate and push a new image when relevant, and to update the documentation workflow with the new tag and digest.

For Pull Requests from branches of the cilium/cilium repository, we can add a new commit directly to update the image. For branches from external forks, we require an additional step from the PR author to update the image reference.

Fixes: #24019

Status: Currently a draft, because we're missing scripts and docs regarding the update for PRs with branches on external forks. Also, nothing has been tested yet.

qmonnet committed:
ci: Automate generation and update of docs-builder image

The docs-builder image is used for the CI wofkflow for testing
documentation. Each time we change the list of requirements (Sphinx
dependencies etc.) or the Dockerfile for that image, we have to generate
a new image to use in the CI workflow.

So far these images have been generated and pushed to Docker manually
(thanks Joe!). But this is not fun or practical [citation needed]. This
commit introduces a new CI workflow to generate and push a new image
when relevant, and to update the documentation workflow with the new tag
and digest.

For Pull Requests from branches of the cilium/cilium repository, we can
add a new commit directly to update the image. For branches from
external forks, we require an additional step from the PR author to
update the image reference.

Related: #24019
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

joestringer commented on .github/workflows/build-images-docs-builder.yaml:
```diff
@@ -0,0 +1,142 @@
+        uses: docker/login-action@f4ef78c080cd8ba55a85445d5b36e214a81df20a # v2.1.0
+        with:
+          username: ${{ secrets.DOCKER_HUB_RELEASE_USERNAME }}
+          password: ${{ secrets.DOCKER_HUB_RELEASE_PASSWORD }}
```
TODO(@joestringer): Make sure these are configured and have the right level of access rights.

joestringer commented on .github/workflows/build-images-docs-builder.yaml:
```diff
@@ -0,0 +1,142 @@
+          mkdir -p image-digest/
+          echo "## docs-builder" > image-digest/docs-builder.txt
+          echo "" >> image-digest/docs-builder.txt
+          echo "\`docker://${{ env.QUAY_ORGANIZATION_DEV }}/cilium-runtime:${{ steps.docs-builder-tag.outputs.tag }}@${{ steps.docker_build_docs_builder.outputs.digest }}\`" >> image-digest/docs-builder.txt
```
cilium-runtime?

aanm commented on .github/workflows/build-images-docs-builder.yaml:
```diff
@@ -0,0 +1,143 @@
+
+# Any change in triggers needs to be reflected in the concurrency group.
+on:
+  pull_request:
```
If we are ok with this, this will only work for PRs opened from a branch in this repository.

qmonnet replied:
This is part of the second commit:
![image](https://user-images.githubusercontent.com/17001771/222743122-f1edf345-e29f-458e-9830-567eb1ac8bf5.png)

I intend to remove it after I manage to test the workflow successfully

qmonnet committed:
ci: Switch to quay.io for docs-builder image (documentation workflow)

We're switching to quay.io to host the docs-builder images. Among other
things, this switch brings the possibility to automatically update the
image from PRs on the Cilium repository, when relevant files are
updated.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

aanm reviewed (commented):
cc @qmonnet the secrets are now added

joestringer commented on .github/workflows/build-images-docs-builder.yaml:
```diff
@@ -0,0 +1,147 @@
+        with:
+          provenance: false
+          context: ./Documentation
+          file: ./Documentation/Dockerfile
```
Is there a way to edit this Dockerfile to expose environment variables from the host and echo them to the GHA output?

joestringer replied:
I guess with `environment: docs-builder` this requires docs-structure review anyway so as long as docs reviewers are aware that they need to check for env var abuse in any changes that are proposed against these files, then that is fine. Otherwise if docs-structure reviewers can't be trusted to enforce this then we'd need to think about how to ensure that folks with the appropriate knowledge are put in the path for running these checks.

joestringer replied:
I also wasn't 100% sure whether the secrets are available if a PR is opened from an external fork. If it requires push access to cilium/cilium then that is also a mitigating factor, as that requires org-member privileges IIRC.

qmonnet replied:
> Is there a way to edit this Dockerfile to expose environment variables from the host and echo them to the GHA output?

Excellent question. As far as I'm aware, the only way would be to use the `ARG` directive in the Dockerfile, _**and**_ the `--build-arg` argument for the `docker` invocation; but the workflow file does not use the relevant `build-args` parameter for the `docker/build-push-action` action, so it shouldn't be possible to do it this way.

This being said, I'm not well versed on Docker security issues, so there could be something else I'm not aware of.

nbusseneau replied:
This is correct. FYI we have the exact same setup in the image build workflows. These workflows run on `pull_request_target` so they run in a privileged context, and are indeed able to access repository secrets (and in general they have various dangerous permissions), so we must be careful as to what comes from the PR and how we use it.

In the case of `docker/build-push-action` using a Dockerfile coming from the PR it is fine because as said by Quentin, for the Dockerfile to be able to access the environment variables from the privileged workflow we would need to pass the secrets explicitly to the Dockerfile (as is done here: https://docs.docker.com/build/ci/github-actions/secrets/).

joestringer reviewed (commented):
Just one question, but I'll encourage additional scrutiny from other @cilium/github-sec members given the use of `pull_request_target` here.

learnitall commented on .github/workflows/build-images-docs-builder.yaml:
```diff
@@ -0,0 +1,147 @@
+      - name: Update docs-builder image reference in CI workflow
+        if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
+        run: |
+          Documentation/update-docs-builder-image.sh "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}@${{ steps.docker_build_docs_builder.outputs.digest }}"
```
Wouldn't this run the not trusted `update-docs-builder-image.sh` from the pull request branch, rather than the default trusted branch? If so, a user could change this script and run whatever they wanted in the workflow.

joestringer replied:
Note that triggering the overall workflow is only done with @cilium/docs-structure approval due to `environment: docs-builder` plus the cilium/cilium configuration for that environment. So we are partially relying here on those reviewers doing due diligence on any PRs requesting to change build infrastructure.

It would probably be better practice to do some combination of the following:

- Log out of Quay after images are pushed
- Directly call docker from the workflow to wrap & run the docs script

learnitall replied:
Ah ok that makes sense. Should this be non-blocking? Those options sound good. I think we could also save a copy of the trusted script in the beginning of the job and use it later o?

qmonnet replied:
Both proposals look good, but I'm not sure how to implement them.

> * Log out of Quay after images are pushed

I don't know how to explicitly logout from Quay. The Action we use supports a `logout` parameter, already enabled by default, but logging out occurs only at the end of the job. One solution I see would be to split the current job into several ones, starting a new one after `Build docs-builder image` (which I think is the only step that requires the log in)?

> * Directly call docker from the workflow to wrap & run the docs script

Do you mean replacing the script invocation with the following?"

```console
$ docker run --rm -v $PWD:/cilium -w /cilium cilium/docs-builder \
    bash -c "git config --global --add safe.directory /cilium && /cilium/Documentation/update-docs-builder-image.sh 'quay.io/...'"
```

aanm replied:
> One solution I see would be to split the current job into several ones, starting a new one after Build docs-builder image (which I think is the only step that requires the log in)?

I would also be in favor of this approach.

learnitall reviewed (changes requested):
Quick question in regards to the `update-docs-builder-image.sh` before we merge.

learnitall commented on .github/workflows/build-images-docs-builder.yaml:
```diff
@@ -0,0 +1,196 @@
+        with:
+          path: image-digest/
+
+      - name: Output image digest
+        shell: bash
+        run: |
+          cd image-digest/
+          find -type f | sort | xargs -d '\n' cat
```
Could this step be merged into the previous job?

qmonnet replied:
:thinking: Let me see, I didn't pay too much attention to that step, but it seems that in the workflow I copied from (base image generation), we needed a separate job to retrieve images from the different jobs spawned from the matrix; and in our case, we don't need that, given that we only build one image. Is this what you mean?

I'll give it a try

qmonnet replied:
OK, seems like [it's still working](https://github.com/cilium/cilium/actions/runs/5799571701/job/15719921971)

<details><summary>Incremental diff</summary>

``` diff
diff --git a/.github/workflows/build-images-docs-builder.yaml b/.github/workflows/build-images-docs-builder.yaml
index b4157a621f0d..fa48b23aa73e 100644
--- a/.github/workflows/build-images-docs-builder.yaml
+++ b/.github/workflows/build-images-docs-builder.yaml
@@ -141,8 +141,8 @@ jobs:
           git diff HEAD^
           git push https://x-access-token:${{ steps.get_token.outputs.app_token }}@github.com/${{ env.QUAY_ORGANIZATION }}/cilium.git HEAD:"$REF"
 
-  retrieve-digest:
-    name: Retrieve and upload image digest
+  image-digest:
+    name: Retrieve and display image digest
     needs: build-and-push
     if: needs.build-and-push.outputs.digest
     runs-on: ubuntu-20.04
@@ -174,21 +174,6 @@ jobs:
           path: image-digest
           retention-days: 1
 
-  image-digests:
-    name: Display digests
-    needs: retrieve-digest
-    runs-on: ubuntu-20.04
-    steps:
-      - name: Create image digest directory
-        shell: bash
-        run: |
-          mkdir -p image-digest/
-
-      - name: Download image digest
-        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
-        with:
-          path: image-digest/
-
       - name: Output image digest
         shell: bash
         run: |
```
</details>

learnitall replied:
Yep that's what I was getting at! The incremental diff looks great!

learnitall reviewed (commented):
Awesome! One quick question about the last job, but LGTM.

Commit:
ci: update docs-builder

Signed-off-by: Cilium Imagebot <noreply@cilium.io>

learnitall reviewed (approved):
LGTM! :partying_face:

aanm commented on .github/workflows/build-images-docs-builder.yaml:
```diff
@@ -0,0 +1,181 @@
+
+# Any change in triggers needs to be reflected in the concurrency group.
+on:
+  pull_request:
```
If we are back to pull_request target there's no need tos split it by jobs since PRs opened from forks won't have access to secrets.

qmonnet replied:
:warning: We're not! This is just a test commit so I could validate the workflow
